src/software/ai/hl/stp/play_selection_fsm.h

```diff
@@ -0,0 +1,141 @@
+        GameState game_state;
+    };
+
+    /**
+     * Creates a play selection FSM
+     *
+     * @param ai_config the play config for this play fsm
+     */
+    explicit PlaySelectionFSM(std::shared_ptr<const AiConfig> ai_config);
+
+    /**
+     * Guards for whether the game state is stopped, halted, playing, or in set up
+     *
+     * @param event The PlaySelection::Update event
+     *
+     * @return whether the gamestate is stopped, halted, playing, or in set up
+     */
+    bool gameStateStopped(const Update& event);
+    bool gameStateHalted(const Update& event);
+    bool gameStatePlaying(const Update& event);
+    bool gameStateSetup(const Update& event);
+
+    /**
+     * Guards for whether the play is being overridden
+     *
+     * @param event The PlaySelection::Update event
+     *
+     * @return whether the play is being overridden
+     */
+    bool playOverridden(const Update& event);
+
+    /**
+     * Action to set up the OverridePlay, SetPlay, StopPlay, HaltPlay, or OffensePlay
+     *
+     * @param event The PlaySelection::Update event
+     *
+     */
+    void setupOverridePlay(const Update& event);
+    void setupSetPlay(const Update& event);
+    void setupStopPlay(const Update& event);
+    void setupHaltPlay(const Update& event);
+    void setupOffensePlay(const Update& event);
+
+    auto operator()()
+    {
+        using namespace boost::sml;
+
+        DEFINE_SML_STATE(OverridePlay)
+        DEFINE_SML_STATE(SetPlay)
+        DEFINE_SML_STATE(Halt)
+        DEFINE_SML_STATE(Playing)
+        DEFINE_SML_STATE(Stop)
+
+        DEFINE_SML_GUARD(playOverridden)
+        DEFINE_SML_GUARD(gameStateStopped)
+        DEFINE_SML_GUARD(gameStateHalted)
+        DEFINE_SML_GUARD(gameStatePlaying)
+        DEFINE_SML_GUARD(gameStateSetup)
+
+        DEFINE_SML_EVENT(Update)
+
+        DEFINE_SML_ACTION(setupOverridePlay)
+        DEFINE_SML_ACTION(setupSetPlay)
+        DEFINE_SML_ACTION(setupStopPlay)
+        DEFINE_SML_ACTION(setupHaltPlay)
+        DEFINE_SML_ACTION(setupOffensePlay)
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
+
+            // Check for transitions to other states, if not then default to running the
+            // current play
+            *Halt_S + Update_E[gameStateStopped_G] / setupStopPlay_A   = Stop_S,
+            Halt_S + Update_E[gameStatePlaying_G] / setupOffensePlay_A = Playing_S,
+            Halt_S + Update_E[gameStateSetup_G] / setupSetPlay_A       = SetPlay_S,
+            Halt_S + Update_E[playOverridden_G] / setupOverridePlay_A  = OverridePlay_S,
+
+            // If play overridden then continue running the overridden play, otherwise
+            // check game state
+            OverridePlay_S + Update_E[playOverridden_G] = OverridePlay_S,
+            OverridePlay_S + Update_E[gameStateHalted_G] / setupHaltPlay_A = Halt_S,
+            OverridePlay_S + Update_E[gameStatePlaying_G] / setupOffensePlay_A =
```

Comment:
what happened to DefencePlay

Reply:
it will be handled by https://github.com/UBC-Thunderbots/Software/issues/2440